src/www/system_certmanager.php

```diff
@@ -84,6 +84,38 @@ function csr_get_modulus($str_crt, $decode = true)
     return cert_get_modulus($str_crt, $decode, 'csr');
 }
 
+function parse_csr($csr)
+{
+    // TODO: very bad implementation and should be rewritten before merging (just for testing now)
+    // TODO: implement other extensions
+
+    $ret = array();
+    $ret['parse_success'] = true;
+    $ret['subject'] = openssl_csr_get_subject($csr);
+    if ($ret['subject'] === false) {
+        return array('parse_success' => false);
+    }
+
+    $process = proc_open('/usr/local/bin/openssl req -text -noout | grep "X509v3 Subject Alternative Name:" -A 1 | tail -n 1 | sed -e "s/^ *//g"', array(array("pipe", "r"), array("pipe", "w"), array("pipe", "w")), $pipes);
```

Comment:
This seems like a very ugly solution attempt to me.

I like the solution from https://gist.github.com/dol/e9f3d682529ae7ee368f0e6862e16a87#file-cert-php-L58 much better. 

It uses https://github.com/sop/x509 which in turn depends on a few other libraries. As OPNsense is using X.509 certificates in many places, I'd believe it could be a good idea to add these libraries, but that's more of a question to the core developers.

Reply:
Yes, exactly. This is why I commented TODO lines that I will do beforre merging some lines above.

I'm currently surveying how to install some php modules from ports (it seems I also have to change opnsense/ports repository?) and install sop/x509, sop/crypto-tools on vendor directory.

After finding out some set of tasks to prepare environment, I will comment abount it.

Reply:
Precisely: not just very ugly but it does not work correctly (at least if CSR has IP address in SAN)

Reply:
@nhirokinet @wschlich I'm not exactly sure what we're trying to do here (haven't read all the code), but php has quite some x509 functions built in natively nowadays.  We're always a bit cautious in adding new dependencies to core. If you can paste a simple example of the issue your trying to solve, maybe I can help searching some simple solution.

Reply:
@AdSchellevis (At least in my understanding, )

Tasks:

- Sign a CSR
  - Some x509 extensions, which can be included in CSR and certificate, like SANs, are standard and are a must today.

Problem for the task:

- In OpenSSL design, callers must explicitly designate what x509 extensions, like SAN, to be signed. It is same in PHP's x509 functions built, which relies on OpenSSL.
  - [@wschlich commented for detail](https://github.com/opnsense/core/pull/3234#issuecomment-464081639)
- This means that the PHP code (or other method who creates openssl.cnf) must know x509 extensions values, written in CSR.
  - Difference from previous system_certmanager.php here is `written in CSR`. Current "Create an Internal Certificate" does not have to CSR because CSR is just created in same function.
- Despite there are a lot of PHP functions to parse x509 certificates (most starting with `openssl_x509_`), there seems to be a little for CSR (most starting with `openssl_csr_`), none of them seem to be able to parse x509 extension.
  - [PHP: OpenSSL Functions - Manual](http://php.net/manual/en/ref.openssl.php)

Reply:
@nhirokinet ok, if there's no glue for this csr task in php, that's a bummer. sop/x509 at first glance seems to be a rather large and complex module, but if there are no better alternatives, we might need to consider it.